staging/src/k8s.io/client-go/rest/transport.go

```diff
@@ -118,3 +128,45 @@ func (c *Config) TransportConfig() (*transport.Config, error) {
 func (c *Config) Wrap(fn transport.WrapperFunc) {
 	c.WrapTransport = transport.Wrappers(c.WrapTransport, fn)
 }
+
+var CertificateCacheExpiryPeriod = time.Second
+
+type certificateCacheEntry struct {
+	*tls.Certificate
+	error
+	time.Time
+}
+
+func (c *certificateCacheEntry) isExpired() bool {
+	return time.Now().Sub(c.Time) > CertificateCacheExpiryPeriod
+}
+
+func newCertificateCacheEntry(certFile, keyFile string) certificateCacheEntry {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	return certificateCacheEntry{&cert, err, time.Now()}
+}
+
+// cachingCertificateLoader ensures that we don't hammer the filesystem when opening many connections
+// the underlying cert files are read at most once every second
+func cachingCertificateLoader(certFile, keyFile string) func() (*tls.Certificate, error) {
+	current := newCertificateCacheEntry(certFile, keyFile)
+	var currentMtx sync.RWMutex
+
+	return func() (*tls.Certificate, error) {
+		currentMtx.RLock()
+		if current.isExpired() {
+			currentMtx.RUnlock()
+
+			currentMtx.Lock()
```

Comment:
I expected to see this gotten before the readunlock

Reply:
Won't that cause deadlock? I can't get a write lock until the read is unlocked

Reply:
and I've forgotten just how bad the golang sync libraries are. :(